design/external-authorization-design.md

```diff
@@ -0,0 +1,453 @@
+# External Authorization Support
+
+Status: Reviewing
+
+## Abstract
+This document describes a design for performing request authorization for virtual hosts hosted by Contour.
+
+## Background
+
+- [#432][5] Provide support for integrating with an external auth service.
+- [#2459][3] External auth background and goals.
+
+## Goals
+- Allow operators to integrate existing authorization services with Contour.
+- Support arbitrary types of authentication and authorization.
+- Decouple Contour from external authorization, so it can evolve at an independent rate.
+- Integrate cleanly with the HTTPProxy API.
+
+## Non Goals
+
+- Support for multiple external authorization mechanisms.
```

Comment:
Can you clarify these two bits? I'm not sure what is not going to be a goal.

Reply:
What I meant by this is that different proxies have different protocol mechanisms for incorporating external authorization servers. We aren't going to support multiple mechanisms (i.e. you can't use your external auth server from nginx unmodified).

Reply:
Sorry still not 100% following you on this thought. Could you rewrite the nongoal in a way that better explains it? Right now it reads as you aren't going to support the auth provider that I want to bring which I know is not the case.

Reply:
> Support for multiple external authorization mechanisms (i.e. the protocol the proxy uses to communicate with the authorization server).

The mechanism is the protocol by which the proxy communicates with the authorization server.

> Various proxies have different mechanisms for integrating external services, but it is impractical to support them all.

Nginx, envoy, traefik, etc all use different external authorization mechanisms (ie. communicating with the authorization server follows different protocols or conventions). What I'm getting at is that it is not a goal for an authorization server written for nginx to work with Contour.

Reply:
I guess the ask here was to make the doc more clear to remove this confusion.